### ISSUE TITLE
thread sanitizer seems false positive on Linux x64

### ISSUE THREAD

**Issue by zchrissirhcz** [llvm/llvm-project#53241]
## Problem Description
My C++ project use thread sanitizer for data race inspection. There is one reported data race with latest clang compiler, seems to be false positive (or true positive, due to my poor experience in MT).

The reported data race is about a variable  `s_in_parallel`:
```c++
// my.cpp
namespace my {

static int s_in_parallel = 0; 
```

The reported write of this variable is:
```c++
// my.cpp, line 75
if (MY_XADD(&s_in_parallel, 1) == 0)
```

The reporated read of this variable is:
```c++
// my.cpp, line 98
return s_in_parallel == 0 ? getNumThreads() : 1;
```

**I thought the write operation of `s_in_parallel`, MY_XADD(), is already atomic operation, thus the reading of `s_in_parallel` doesn't require to be atomic. But tsan reports it as data race, which makes confuse.**.

The error message:
```
num_threads = 1
num_threads = 2
s_max_num_threads is 1
s_max_num_threads is 2
==================
WARNING: ThreadSanitizer: data race (pid=1502005)
  Atomic write of size 4 at 0x0000018f66ec by thread T2:
    #0 my::parallel_for_(int, int, void (* const&)(int, void*), void*) /home/zz/work/test/thread/my.cpp:75:9 (testbed+0x4cde1c)
    #1 void my::parallel_for<int, my_test_parallel_MultipleThreads()::$_1::operator()() const::'lambda'(int)>(int, int, my_test_parallel_MultipleThreads()::$_1::operator()() const::'lambda'(int) const&) /home/zz/work/test/thread/my.hpp:66:9 (testbed+0x4ce5a1)
    #2 my_test_parallel_MultipleThreads()::$_1::operator()() const /home/zz/work/test/thread/testbed.cpp:26:9 (testbed+0x4ce5a1)
    #3 void std::__invoke_impl<void, my_test_parallel_MultipleThreads()::$_1>(std::__invoke_other, my_test_parallel_MultipleThreads()::$_1&&) /usr/lib/gcc/x86_64-linux-gnu/9/../../../../include/c++/9/bits/invoke.h:60:14 (testbed+0x4ce5a1)
    #4 std::__invoke_result<my_test_parallel_MultipleThreads()::$_1>::type std::__invoke<my_test_parallel_MultipleThreads()::$_1>(my_test_parallel_MultipleThreads()::$_1&&) /usr/lib/gcc/x86_64-linux-gnu/9/../../../../include/c++/9/bits/invoke.h:95:14 (testbed+0x4ce5a1)
    #5 void std::thread::_Invoker<std::tuple<my_test_parallel_MultipleThreads()::$_1> >::_M_invoke<0ul>(std::_Index_tuple<0ul>) /usr/lib/gcc/x86_64-linux-gnu/9/../../../../include/c++/9/thread:244:13 (testbed+0x4ce5a1)
    #6 std::thread::_Invoker<std::tuple<my_test_parallel_MultipleThreads()::$_1> >::operator()() /usr/lib/gcc/x86_64-linux-gnu/9/../../../../include/c++/9/thread:251:11 (testbed+0x4ce5a1)
    #7 std::thread::_State_impl<std::thread::_Invoker<std::tuple<my_test_parallel_MultipleThreads()::$_1> > >::_M_run() /usr/lib/gcc/x86_64-linux-gnu/9/../../../../include/c++/9/thread:195:13 (testbed+0x4ce5a1)
    #8 <null> <null> (libstdc++.so.6+0xd6d83) (BuildId: c45813107d1487f8fc7197424aa6f1e0be00fe6c)

  Previous read of size 4 at 0x0000018f66ec by thread T1:
    #0 my::getAvailableNumThreads() /home/zz/work/test/thread/my.cpp:98:12 (testbed+0x4cdf48)
    #1 void my::parallel_for<int, my_test_parallel_MultipleThreads()::$_0::operator()() const::'lambda'(int)>(int, int, my_test_parallel_MultipleThreads()::$_0::operator()() const::'lambda'(int) const&) /home/zz/work/test/thread/my.hpp:64:29 (testbed+0x4ce4a5)
    #2 my_test_parallel_MultipleThreads()::$_0::operator()() const /home/zz/work/test/thread/testbed.cpp:15:9 (testbed+0x4ce4a5)
    #3 void std::__invoke_impl<void, my_test_parallel_MultipleThreads()::$_0>(std::__invoke_other, my_test_parallel_MultipleThreads()::$_0&&) /usr/lib/gcc/x86_64-linux-gnu/9/../../../../include/c++/9/bits/invoke.h:60:14 (testbed+0x4ce4a5)
    #4 std::__invoke_result<my_test_parallel_MultipleThreads()::$_0>::type std::__invoke<my_test_parallel_MultipleThreads()::$_0>(my_test_parallel_MultipleThreads()::$_0&&) /usr/lib/gcc/x86_64-linux-gnu/9/../../../../include/c++/9/bits/invoke.h:95:14 (testbed+0x4ce4a5)
    #5 void std::thread::_Invoker<std::tuple<my_test_parallel_MultipleThreads()::$_0> >::_M_invoke<0ul>(std::_Index_tuple<0ul>) /usr/lib/gcc/x86_64-linux-gnu/9/../../../../include/c++/9/thread:244:13 (testbed+0x4ce4a5)
    #6 std::thread::_Invoker<std::tuple<my_test_parallel_MultipleThreads()::$_0> >::operator()() /usr/lib/gcc/x86_64-linux-gnu/9/../../../../include/c++/9/thread:251:11 (testbed+0x4ce4a5)
    #7 std::thread::_State_impl<std::thread::_Invoker<std::tuple<my_test_parallel_MultipleThreads()::$_0> > >::_M_run() /usr/lib/gcc/x86_64-linux-gnu/9/../../../../include/c++/9/thread:195:13 (testbed+0x4ce4a5)
    #8 <null> <null> (libstdc++.so.6+0xd6d83) (BuildId: c45813107d1487f8fc7197424aa6f1e0be00fe6c)

  Location is global 'my::s_in_parallel' of size 4 at 0x0000018f66ec (testbed+0x18f66ec)

  Thread T2 (tid=1502008, running) created by main thread at:
    #0 pthread_create /home/zz/work/github/llvm-project/compiler-rt/lib/tsan/rtl/tsan_interceptors_posix.cpp:1017:3 (testbed+0x44d1dd)
    #1 std::thread::_M_start_thread(std::unique_ptr<std::thread::_State, std::default_delete<std::thread::_State> >, void (*)()) <null> (libstdc++.so.6+0xd7048) (BuildId: c45813107d1487f8fc7197424aa6f1e0be00fe6c)
    #2 main /home/zz/work/test/thread/testbed.cpp:37:5 (testbed+0x4ce403)

  Thread T1 (tid=1502007, finished) created by main thread at:
    #0 pthread_create /home/zz/work/github/llvm-project/compiler-rt/lib/tsan/rtl/tsan_interceptors_posix.cpp:1017:3 (testbed+0x44d1dd)
    #1 std::thread::_M_start_thread(std::unique_ptr<std::thread::_State, std::default_delete<std::thread::_State> >, void (*)()) <null> (libstdc++.so.6+0xd7048) (BuildId: c45813107d1487f8fc7197424aa6f1e0be00fe6c)
    #2 main /home/zz/work/test/thread/testbed.cpp:37:5 (testbed+0x4ce403)

SUMMARY: ThreadSanitizer: data race /home/zz/work/test/thread/my.cpp:75:9 in my::parallel_for_(int, int, void (* const&)(int, void*), void*)
==================
ThreadSanitizer: reported 1 warnings
```

## Reproduce
OS: ubuntu 20.04

Compile latest clang compiler:
```
(base) zz@machine% clang --version
clang version 14.0.0 (/home/zz/work/github/llvm-project/clang cc35161dc716ab5b0e66b15ac320ebd451a2bc44)
Target: x86_64-unknown-linux-gnu
Thread model: posix
InstalledDir: /home/zz/soft/llvm14-dev/bin
```

### Full code files:
4 files:
- my.hpp
- my.cpp
- testbed.cpp
- CMakeLists.txt

my.hpp:
```c++
#pragma once

// copied from OpenCV:  https://github.com/opencv/opencv/blob/4.x/modules/core/include/opencv2/core/cvdef.h

/****************************************************************************************\
*          exchange-add operation for atomic operations on reference counters            *
\****************************************************************************************/

#ifdef MY_XADD
  // allow to use user-defined macro
#elif defined __GNUC__ || defined __clang__
#  if defined __clang__ && __clang_major__ >= 3 && !defined __ANDROID__ && !defined __EMSCRIPTEN__ && !defined(__CUDACC__)  && !defined __INTEL_COMPILER
#    ifdef __ATOMIC_ACQ_REL
#      define MY_XADD(addr, delta) __c11_atomic_fetch_add((_Atomic(int)*)(addr), delta, __ATOMIC_ACQ_REL)
#    else
#      define MY_XADD(addr, delta) __atomic_fetch_add((_Atomic(int)*)(addr), delta, 4)
#    endif
#  else
#    if defined __ATOMIC_ACQ_REL && !defined __clang__
       // version for gcc >= 4.7
#      define MY_XADD(addr, delta) (int)__atomic_fetch_add((unsigned*)(addr), (unsigned)(delta), __ATOMIC_ACQ_REL)
#    else
#      define MY_XADD(addr, delta) (int)__sync_fetch_and_add((unsigned*)(addr), (unsigned)(delta))
#    endif
#  endif
#elif defined _MSC_VER && !defined RC_INVOKED
#  include <intrin.h>
#  define MY_XADD(addr, delta) (int)_InterlockedExchangeAdd((long volatile*)addr, delta)
#else
  #ifdef OPENCV_FORCE_UNSAFE_XADD
    CV_INLINE int MY_XADD(int* addr, int delta) { int tmp = *addr; *addr += delta; return tmp; }
  #else
    #error "OpenCV: can't define safe MY_XADD macro for current platform (unsupported). Define MY_XADD macro through custom port header (see OPENCV_INCLUDE_PORT_FILE)"
  #endif
#endif





#define USE_PARALLEL_FOR 1
#define HAVE_THREAD_PF 1
#define HAVE_PARALLEL_FRAMEWORK 1

namespace my {

int getAvailableNumThreads();

void setNumThreads(int num_threads);
int getNumThreads();
int getDefaultNumThreads();
int getNumberOfCPUs();
int getMaxNumThreads();

typedef void (*ParallelLoopFunc)(int index, void* param);
void parallel_for_(int first, int last, const ParallelLoopFunc& func, void* param=0);


template<typename IndexType, typename Function> inline
void parallel_for(IndexType first, IndexType last, const Function& func)
{
#if USE_PARALLEL_FOR
    // fast return
    if (last - first > 1 && my::getAvailableNumThreads() > 1)
    {
        my::parallel_for_((int) first, (int) last, [](int index, void* param) {
            (*static_cast<Function*>(param))((IndexType) index);
        }, (void*) &func);
    }
    else
#endif // USE_PARALLEL_FOR
    {
        for (IndexType i = first; i < last; ++i)
            func(i);
    }
}

} // namespace my
```

my.cpp
```c++
#include "my.hpp"

//#include <toy/context.private.hpp>
#include <mutex>
#include <unistd.h> // sysconf()

namespace my {

static int s_in_parallel = 0;
static thread_local int s_num_threads = 0;

#define __MY_CAT__(x, y) x ## y
#define MY_CAT(x, y) __MY_CAT__(x, y)

#define MY_MUTEX(var, ...) __VA_ARGS__ std::mutex var
#define MY_MUTEX_LOCK(var) std::lock_guard<std::mutex> MY_CAT(lock_, __COUNTER__)(var)

MY_MUTEX(s_num_threads_mtx, static);
static int s_max_num_threads = 0;

template<class T> T clamp(const T& x, const T& a, const T& b) { return x < a ? a : x > b ? b : x; }

void setNumThreads(int threads)
{
    threads = clamp(threads >= 0 ? threads : my::getDefaultNumThreads(), 1, my::getMaxNumThreads());
    s_num_threads = threads;

    MY_MUTEX_LOCK(s_num_threads_mtx);

    if (threads <= s_max_num_threads)
        return;

    s_max_num_threads = threads;
    fprintf(stderr, "s_max_num_threads is %d\n", s_max_num_threads);
}


inline int getNumberOfCPUsImpl()
{
    // http://man7.org/linux/man-pages/man3/sysconf.3.html
    // The number of processors currently online (available)
    return (int) sysconf(_SC_NPROCESSORS_ONLN);
}


inline int getDefaultNumThreadsImpl()
{
    const char* str = getenv("MY_NUM_THREADS");
    if (str)
    {
        char* end;
        const int num_threads = (int) strtol(str, &end, 10);
        if (end > str)
            return std::min(1, num_threads);
    }

    return std::max(1, getNumberOfCPUs());
}


} // namespace my


void my::parallel_for_(int first, int last, const ParallelLoopFunc& func, void* param)
{
    // Multi-threads is not needed.
    if (s_num_threads <= 1 || last - first <= 1)
    {
        for (int i = first; i < last; ++i)
            func(i, param);
        return;
    }

    // Check and increase s_in_parallel in single atomic operation.
    if (MY_XADD(&s_in_parallel, 1) == 0)
    {
#if HAVE_THREAD_PF

        const int num_threads = s_num_threads >= 0 ? s_num_threads : my::getDefaultNumThreads();
        //tv::Context::instance().getPTM()->run(num_threads, first, last, func, param);

#endif // HAVE_{}

        MY_XADD(&s_in_parallel, -1);
    }
    else
    {
        MY_XADD(&s_in_parallel, -1);

        for (int i = first; i < last; ++i)
            func(i, param);
    }
}


int my::getAvailableNumThreads()
{
    return s_in_parallel == 0 ? getNumThreads() : 1;   // old, cause data race
    //return MY_XADD(&s_in_parallel, 0) == 0 ? getNumThreads() : 1; // new. No data race.
}


int my::getNumThreads(void)
{
    if (s_num_threads > 0)
        return s_num_threads;

#if HAVE_THREAD_PF

    return getDefaultNumThreads();

#else // HAVE_{}

    return 1;

#endif // HAVE_{}
}


int my::getDefaultNumThreads()
{
    static const int default_num_threads = getDefaultNumThreadsImpl();
    return default_num_threads;
}

int my::getNumberOfCPUs()
{
    static int ncpus = getNumberOfCPUsImpl();
    return ncpus;
}


int my::getMaxNumThreads()
{
#if HAVE_PARALLEL_FRAMEWORK
    return getNumberOfCPUs();
#endif
    return -1;
}
```

testbed.cpp
```c++
#include <gtest/gtest.h>
#include <thread>
#include "my.hpp"

void my_test_parallel_MultipleThreads()
{
    std::thread ts[2];

    int num_threads = 1;
    fprintf(stderr, "num_threads = %d\n", num_threads);
    ts[0] = std::thread([num_threads] {
        const int N = 2;
        int count = 0;
        my::setNumThreads(num_threads);
        my::parallel_for(0, N, [&count](int) {

        });
    });

    num_threads = 2;
    fprintf(stderr, "num_threads = %d\n", num_threads);
    ts[1] = std::thread([num_threads] {
        const int N = 2;
        int count = 0;
        my::setNumThreads(num_threads);
        my::parallel_for(0, N, [&count](int) {

        });
    });

    ts[0].join();
    ts[1].join();
}

int main()
{
    my_test_parallel_MultipleThreads();
    return 0;
}
```

CMakeLists.txt
```c++
cmake_minimum_required(VERSION 3.20)
project(hello)

set(CMAKE_CXX_STANDARD 11)
set(CMAKE_EXPORT_COMPILE_COMMANDS ON)

set(CMAKE_BUILD_TYPE Release)

#tsan
option(USE_TSAN "Use Thread Sanitizer?" ON)
#--------------------------------------------------
# globally setting
#--------------------------------------------------
# https://stackoverflow.com/a/65019152/2999096
if(USE_TSAN)
  if (MSVC)
    message(WARNING "ThreadSanitizer for MSVC must be enabled manually in project's property page.")
  else()
    message(STATUS ">>> USE_TSAN: YES")
    message(STATUS ">>> CMAKE_BUILD_TYPE: ${CMAKE_BUILD_TYPE}")
    if (CMAKE_BUILD_TYPE MATCHES "Debug")
      message(STATUS ">>>!!!!!!!!!!!Debug mode")
      # On debug mode
      set(CMAKE_C_FLAGS_DEBUG "${CMAKE_C_FLAGS_DEBUG} -fno-omit-frame-pointer -fsanitize=thread")
      set(CMAKE_CXX_FLAGS_DEBUG "${CMAKE_CXX_FLAGS_DEBUG} -fno-omit-frame-pointer -fsanitize=thread")
      set(CMAKE_LINKER_FLAGS_DEBUG "${CMAKE_LINKER_FLAGS_DEBUG} -fno-omit-frame-pointer -fsanitize=thread")
    elseif (CMAKE_BUILD_TYPE MATCHES "Release")
      # Note: `-g` is explicitly required here, even if NDK android.toolchain.cmake's `-g` exists.
      set(CMAKE_C_FLAGS_RELEASE "${CMAKE_C_FLAGS_RELEASE} -g -fno-omit-frame-pointer -fsanitize=thread")
      set(CMAKE_CXX_FLAGS_RELEASE "${CMAKE_CXX_FLAGS_RELEASE} -g -fno-omit-frame-pointer -fsanitize=thread")
      set(CMAKE_LINKER_FLAGS_RELEASE "${CMAKE_LINKER_FLAGS_RELEASE} -g -fno-omit-frame-pointer -fsanitize=thread")
    elseif (CMAKE_BUILD_TYPE MATCHES "RelWithDebInfo")
      set(CMAKE_C_FLAGS_RELWITHDEBINFO "${CMAKE_C_FLAGS_RELWITHDEBINFO} -g -fno-omit-frame-pointer -fsanitize=thread")
      set(CMAKE_CXX_FLAGS_RELWITHDEBINFO "${CMAKE_CXX_FLAGS_RELWITHDEBINFO} -g -fno-omit-frame-pointer -fsanitize=thread")
      set(CMAKE_LINKER_FLAGS_RELWITHDEBINFO "${CMAKE_LINKER_FLAGS_RELWITHDEBINFO} -g -fno-omit-frame-pointer -fsanitize=thread")
    elseif (CMAKE_BUILD_TYPE MATCHES "MinSizeRel")
      set(CMAKE_C_FLAGS_MINSIZEREL "${CMAKE_C_FLAGS_MINSIZEREL} -g -fno-omit-frame-pointer -fsanitize=thread")
      set(CMAKE_CXX_FLAGS_MINSIZEREL "${CMAKE_CXX_FLAGS_MINSIZEREL} -g -fno-omit-frame-pointer -fsanitize=thread")
      set(CMAKE_LINKER_FLAGS_MINSIZEREL "${CMAKE_LINKER_FLAGS_MINSIZEREL} -g -fno-omit-frame-pointer -fsanitize=thread")
    elseif (CMAKE_BUILD_TYPE EQUAL "None" OR NOT CMAKE_BUILD_TYPE)
      set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -g -fno-omit-frame-pointer -fsanitize=thread")
      set(CMAKE_CXX_FLAGS "${CMAKE_CXX_FLAGS} -g -fno-omit-frame-pointer -fsanitize=thread")
      set(CMAKE_LINKER_FLAGS "${CMAKE_LINKER_FLAGS} -g -fno-omit-frame-pointer -fsanitize=thread")
    endif()
  endif()
else()
  message(STATUS ">>> USE TSan: NO")
endif()


add_executable(testbed
    my.hpp
    my.cpp
    testbed.cpp
)

target_link_libraries(testbed pthread)
```

**Comment by zchrissirhcz** [comment]
I've just also put all the required files to a public repo, with same files.

https://github.com/zchrissirhcz/tsan_test

**Comment by jprotze** [comment]
For atomic memory accesses, the compiler needs to make sure that they are actually seen by other threads. A non-atomic memory access could as well be satisfied by a local copy (e.g., from a register). 
```
while (!s_in_parallel);
```
This might result in an infinite loop, even if another thread is writing to the variable. For the compiler, there is no need to get a new value from iteration to iteration.

Since you use c++11 anyways, just use `static std::atomic<int> s_in_parallel{0};`. This way you are safe and all operations are translated into proper atomic operations.
